src/pil/compiler/powdr_pil.rs

```diff
@@ -0,0 +1,184 @@
+            .collect::<Vec<UUID>>(),
+    );
+
+    // Collect UUIDs of forward signals stored at the circuit level.
+    col_witness.extend(
+        ast.forward_signals
+            .iter()
+            .map(|forward_signal| forward_signal.uuid())
+            .collect::<Vec<UUID>>(),
+    );
+
+    // Collect UUIDs of shared signals stored at the circuit level.
+    col_witness.extend(
+        ast.shared_signals
+            .iter()
+            .map(|shared_signal| shared_signal.uuid())
+            .collect::<Vec<UUID>>(),
+    );
+
+    col_witness
+}
+
+fn collect_step_instances<F, TraceArgs>(
+    ast: &Circuit<F, TraceArgs>,
+    witness: &Option<TraceWitness<F>>,
+) -> HashMap<UUID, Vec<usize>> {
+    let mut step_types_instantiations = HashMap::new();
+
+    if !ast.step_types.is_empty() && witness.is_some() {
+        let step_instances = witness.as_ref().unwrap().step_instances.iter();
+
+        for step_type in ast.step_types.values() {
+            let step_type_instantiation: Vec<usize> = step_instances
+                .clone()
+                .map(|step_instance| {
+                    if step_instance.step_type_uuid == step_type.uuid() {
+                        1
+                    } else {
+                        0
+                    }
+                })
+                .collect();
+            assert_eq!(step_type_instantiation.len(), ast.num_steps);
+            step_types_instantiations.insert(step_type.uuid(), step_type_instantiation);
+        }
+    }
+
+    step_types_instantiations
+}
+
+type ConstraintsByStepID<F> = HashMap<UUID, Vec<Expr<F, Queriable<F>>>>;
+type TransitionsByStepID<F> = HashMap<UUID, Vec<Expr<F, Queriable<F>>>>;
+type LookupsByStepID<F> = HashMap<UUID, Vec<PilLookup<F>>>;
+
+fn compile_steps<F: Clone, TraceArgs>(
+    ast: &Circuit<F, TraceArgs>,
+) -> (
+    ConstraintsByStepID<F>,
+    TransitionsByStepID<F>,
+    LookupsByStepID<F>,
+) {
+    // Group constraints, transitions, and lookups by step type UUID using HashMaps.
+    let mut constraints = HashMap::new();
+    let mut transitions = HashMap::new();
+    let mut lookups = HashMap::new();
+
+    if !ast.step_types.is_empty() {
+        ast.step_types.values().for_each(|step_type| {
+            // Create constraint statements.
+            constraints.insert(
+                step_type.uuid(),
+                step_type
+                    .constraints
+                    .iter()
+                    .map(|constraint| constraint.expr.clone())
+                    .collect::<Vec<Expr<F, Queriable<F>>>>(),
+            );
+
+            transitions.insert(
+                step_type.uuid(),
+                step_type
+                    .transition_constraints
+                    .iter()
+                    .map(|transition| transition.expr.clone())
+                    .collect::<Vec<Expr<F, Queriable<F>>>>(),
+            );
+
+            // Note that there's no lookup table in PIL, so we only need to convert lookups.
```

Comment:
@qwang98 I don't understand this comment, if there is no lookup table in PIL, we should not add them, give a panic if someone tries to compile a circuit with lookup to PIL

Reply:
PIL doesn't have lookup table as to store a set of destination columns. They just store the fixed columns separately. PIL still has lookup (which will be the "application" of the lookup table in the case of Chiquito). I deleted these comments to avoid confusion.